assembly/src/library/mod.rs

```diff
@@ -24,77 +28,412 @@ mod tests;
 // COMPILED LIBRARY
 // ================================================================================================
 
-/// Represents a library where all modules modules were compiled into a [`MastForest`].
+/// Represents a library where all modules were compiled into a [`MastForest`].
+///
+/// A library exports a set of one or more procedures. Currently, all exported procedures belong
+/// to the same top-level namespace.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub struct CompiledLibrary {
+    /// The content hash of this library, formed by hashing the roots of all exports in
+    /// lexicographical order (by digest, not procedure name)
+    digest: RpoDigest,
+    /// A map between procedure paths and the corresponding procedure toots in the MAST forest.
+    /// Multiple paths can map to the same root, and also, some roots may not be associated with
+    /// any paths.
+    exports: BTreeMap<QualifiedProcedureName, Export>,
+    /// The MAST forest underlying this library.
     mast_forest: MastForest,
-    // a path for every `root` in the associated MAST forest
-    exports: Vec<FullyQualifiedProcedureName>,
+}
+
+impl AsRef<CompiledLibrary> for CompiledLibrary {
+    #[inline(always)]
+    fn as_ref(&self) -> &CompiledLibrary {
+        self
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[repr(u8)]
+enum Export {
+    /// The export is contained in the [MastForest] of this library
+    Local(MastNodeId),
+    /// The export is a re-export of an externally-defined procedure from another library
+    External(RpoDigest),
 }
```

Comment:
The main thing I wanted to represent here was whether the export was part of the library, or re-exported from another library. If we make the change you suggest, this information would be obscured. We don't currently expose this information, but I suspect it will come in handy when building packaging infrastructure, so that our tooling can use information like this for analysis purposes.

That said, it's not critical, so I could go either way.

Reply:
I think we should still be able to get this info. For example, we could provide something `CompiledLibrary::get_reexports() -> Vec<(QualifiedProcedureName, RpoDigest)>`. This method would just iterate over roots of the MAST forest and filter out roots which are not direct references to `External` nodes.